src/main/java/org/apache/jasper/servlet/JasperLoader.java

```diff
@@ -130,6 +137,32 @@ public JasperLoader(URL[] urls, ClassLoader parent,
         return findClass(name);
     }
 
+    @Override
+    protected Class<?> findClass(final String name) throws ClassNotFoundException {
+        try {
+            return AccessController.doPrivileged(
+                    new PrivilegedExceptionAction<Class>() {
+                        public Class run() throws ClassNotFoundException {
+                            String path = name.replace('.', '/').concat(".class");
+                            URL res = findResource(path);
+                            if (res != null) {
+                                try (InputStream stream = res.openStream()) {
+                                    int length = stream.available();
```

Comment:
stream.available() only gives an estimate of the length (according to the input stream contract). This will work for files, but if the underlying URL is a remote URL there is no guarantee this will work.